matrix/matrix.go

```diff
@@ -0,0 +1,294 @@
+package matrix
+
+import (
+	"fmt"
+	"image"
+	"math"
+	"log"
+	"runtime"
+)
+
+type Matrix [][]float64
+
+func NewMatrix(row int, col int) Matrix {
+	matrix := make([][]float64, row)
+	for i := range matrix {
+		matrix[i] = make([]float64, col)
+	}
+	return Matrix(matrix)
+}
+
+func deepCopy(m Matrix) Matrix {
+	row := m.Row()
+	col := m.Col()
+	matrix := make([][]float64, row)
+	for i := range matrix {
+		matrix[i] = make([]float64, col)
+		for j := range matrix[i] {
+			matrix[i][j] = m[i][j]
+		}
+	}
+	return Matrix(matrix)
+}
+
+func NewMatrixFromGreyImage(grey *image.Gray) Matrix {
+	bounds := grey.Bounds()
+	dx := bounds.Dx()
+	dy := bounds.Dy()
+
+	matrix := make([][]float64, dx)
+	for i := range matrix {
+		matrix[i] = make([]float64, dy)
+		for j := range matrix[i] {
+			matrix[i][j] = float64(grey.GrayAt(i, j).Y)
+		}
+	}
+
+	return Matrix(matrix)
+}
+
+func Ones(row int, col int) Matrix {
+	matrix := make([][]float64, row)
+	for i := range matrix {
+		matrix[i] = make([]float64, col)
+		for j := range matrix[i] {
+			matrix[i][j] = 1
+		}
+	}
+	return Matrix(matrix)
+}
+
+func Add(a Matrix, matrices ...Matrix) (Matrix, error) {
+	result := deepCopy(a)
```

Comment:
It may not. This function is performing a matrix add operator, if a user only provide one Matrix ( the first param a) here, it will simply return a matrix equal to a, that's what does the deepCopy is about, create another matrix equal to "a" but not "a", which will be used in early return